pkg/reconciler/knativeserving/knativeserving.go

```diff
@@ -168,18 +222,40 @@ func (r *Reconciler) ensureFinalizerRemoval(_ context.Context, _ *mf.Manifest, i
 
 // Delete obsolete resources from previous versions
 func (r *Reconciler) deleteObsoleteResources(ctx context.Context, manifest *mf.Manifest, instance *servingv1alpha1.KnativeServing) error {
-	resources := []*unstructured.Unstructured{
-		// istio-system resources from 0.3.
-		common.NamespacedResource("v1", "Service", "istio-system", "knative-ingressgateway"),
-		common.NamespacedResource("apps/v1", "Deployment", "istio-system", "knative-ingressgateway"),
-		common.NamespacedResource("autoscaling/v1", "HorizontalPodAutoscaler", "istio-system", "knative-ingressgateway"),
-		// config-controller from 0.5
-		common.NamespacedResource("v1", "ConfigMap", instance.GetNamespace(), "config-controller"),
-	}
-	for _, r := range resources {
-		if err := manifest.Client.Delete(r); err != nil {
-			return err
-		}
+	if err := manifest.Delete(); err != nil {
+		return fmt.Errorf("failed to delete the obsolete resources: %w", err)
 	}
 	return nil
 }
+
+func (r *Reconciler) retrieveManifest(ctx context.Context, version string) (mf.Manifest, error) {
+	if val, found := r.manifests[version]; found {
+		return val, nil
+	}
+
+	logger := logging.FromContext(ctx)
+	koDataDir := os.Getenv("KO_DATA_PATH")
+	manifesrDir := fmt.Sprintf("knative-serving/%s", version)
+	manifest, err := mf.NewManifest(filepath.Join(koDataDir, manifesrDir),
+		mf.UseClient(r.mfClient),
+		mf.UseLogger(zapr.NewLogger(logger.Desugar()).WithName("manifestival")))
+
+	if err != nil {
+		return manifest, err
+	}
+
```

Comment:
Here's where you can put your check for an empty manifest and return a new error.

Reply:
Done.